mongodbatlas/fw_provider.go

```diff
@@ -0,0 +1,394 @@
+					stringvalidator.RegexMatches(regexp.MustCompile(`[\w+=,.@:/\-]*`), ""),
+				},
+			},
+			"policy": schema.StringAttribute{
+				Optional:    true,
+				Description: "IAM Policy JSON describing further restricting permissions for the IAM Role being assumed.",
+				Validators: []validator.String{
+					cstmvalidator.StringIsJSON(),
+				},
+			},
+			"policy_arns": schema.SetAttribute{
+				ElementType: types.StringType,
+				Optional:    true,
+				Description: "Amazon Resource Names (ARNs) of IAM Policies describing further restricting permissions for the IAM Role being assumed.",
+			},
+			"role_arn": schema.StringAttribute{
+				Optional:    true,
+				Description: "Amazon Resource Name (ARN) of an IAM Role to assume prior to making API calls.",
+			},
+			"session_name": schema.StringAttribute{
+				Optional:    true,
+				Description: "An identifier for the assumed role session.",
+				Validators: []validator.String{
+					stringvalidator.LengthBetween(2, 64),
+					stringvalidator.RegexMatches(regexp.MustCompile(`[\w+=,.@\-]*`), ""),
+				},
+			},
+			"source_identity": schema.StringAttribute{
+				Optional:    true,
+				Description: "Source identity specified by the principal assuming the role.",
+				Validators: []validator.String{
+					stringvalidator.LengthBetween(2, 64),
+					stringvalidator.RegexMatches(regexp.MustCompile(`[\w+=,.@\-]*`), ""),
+				},
+			},
+			"tags": schema.MapAttribute{
+				ElementType: types.StringType,
+				Optional:    true,
+				Description: "Assume role session tags.",
+			},
+			"transitive_tag_keys": schema.SetAttribute{
+				ElementType: types.StringType,
+				Optional:    true,
+				Description: "Assume role session tag keys to pass to any subsequent sessions.",
+			},
+		},
+	},
+}
+
+func (p *MongodbtlasProvider) Configure(ctx context.Context, req provider.ConfigureRequest, resp *provider.ConfigureResponse) {
+	var (
+		data tfMongodbAtlasProviderModel
+	)
+
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	defaultValuesAndValidations(&data, resp)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	config := Config{
+		PublicKey:    data.PublicKey.ValueString(),
+		PrivateKey:   data.PrivateKey.ValueString(),
+		BaseURL:      data.BaseURL.ValueString(),
+		RealmBaseURL: data.RealmBaseURL.ValueString(),
+	}
+
+	var assumeRoles []tfAssumeRoleModel
+	data.AssumeRole.ElementsAs(ctx, &assumeRoles, true)
+
+	if len(assumeRoles) > 0 {
+		config.AssumeRole = parseTfModel(ctx, &assumeRoles[0])
+		secret := data.SecretName.ValueString()
+		region := data.Region.ValueString()
+		awsAccessKeyID := data.AwsAccessKeyID.ValueString()
+		awsSecretAccessKey := data.AwsSecretAccessKeyID.ValueString()
+		awsSessionToken := data.AwsSessionToken.ValueString()
+		endpoint := data.StsEndpoint.ValueString()
+		var err error
+		config, err = configureCredentialsSTS(&config, secret, region, awsAccessKeyID, awsSecretAccessKey, awsSessionToken, endpoint)
```

Comment:
passing a pointer to a function and then change the value of the object referred by the pointer is not a good patter as decrease readability

Reply:
This is not addressed

Reply:
just to be clear, to avoid any mutations to the parameter that is being passed, the config would be passed as a value and all changes would be present in the returned config correct?

Reply:
no, you can pass config as a reference. My point here is to make sure to not change the object pointed by `config` inside the function

Reply:
From what I understand we have 2 options to address the concern:

- receive the config as value, so it is clear that all changes will only be present in the returned config.
- receive by pointer, and within the function create a new instance of the struct before modifying.

I do feel the first one is more clear, but maybe I am not fully understanding the second option.

Reply:
I am okay with both options. [Keep in mind that option n1 may make linter sad if the struct size is above 80 bytes](https://go-critic.com/overview.html#hugeparam)